Task#2 Container

KroosT opened the pull request:
Type-safe container which can store values of different types.

KroosT committed:
Task #2 Container moved from another branch.

KroosT committed:
Added Util Class Validator.

Contains isArgNotNull() method that checks if passed argument is null and throw exception.

KroosT committed:
Changed @NotNull by @ParametersAreNonnullByDefault.

KroosT committed:
.gitignore changed

vmatelsky commented on Tasks/src/ContainerTask1/Container.java:
```diff
@@ -0,0 +1,46 @@
+
+    Container(final Container<T> container) {
+        Validator.isArgNotNull(container, "container");
+        mList = new ArrayList<>(container.getList());
```
double copying?

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -0,0 +1,65 @@
+        final List<String> listOfString = new ArrayList<>();
+        listOfIntegers.add(30);
+        listOfIntegers.add(28);
+        listOfIntegers.add(1);
```
could be replaced with 
```
final List<Integer> listOfIntegers = Arrays.asList(30, 28, 1);
```

vmatelsky commented on Tasks/src/ContainerTask1/Container.java:
```diff
@@ -0,0 +1,46 @@
+        mList.add(object);
+    }
+
+    T remove(final int index) {
```
use better name to avoid confusion?

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -0,0 +1,65 @@
+    private final Container<String> mContainerOfStrings = new Container<>();
+
+    @Test
+    public void add() throws Exception {
```
test methods should have more specific names

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Test
+    public void remove1() throws Exception {
```
what does `remove1` mean?

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -0,0 +1,65 @@
+        final Container<Integer> integerContainer = new Container<>(mContainerOfIntegers);
+        final Container<String> stringContainer = new Container<>(mContainerOfStrings);
+        Assert.assertNotEquals(integerContainer, mContainerOfIntegers);
+        Assert.assertNotEquals(stringContainer, mContainerOfStrings);
```
what does this test check?
Looks pretty useless

vmatelsky reviewed (changes requested):
Pls fix according to notes

KroosT committed:
Got rid of double copying.

KroosT committed:
Renamed methods to avoid conusion.

KroosT committed:
Got rid of useless tests.

KroosT committed:
Removed last empty line to resolve merge conflict.

vmatelsky commented on Tasks/src/ContainerTask1/Container.java:
```diff
@@ -30,11 +30,11 @@ void add(final T object) {
     }
 
-    void remove(final T object) {
+    void removeObject(final T object) {
```
no need to rename this one: `remove` and `removeByIndex`(or `removeAt`) is ok

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -9,57 +10,31 @@
 public class ContainerTest {
 
     private final Container<Integer> mContainerOfIntegers = new Container<>();
-    private final Container<String> mContainerOfStrings = new Container<>();
```
should be recreated before each test

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -9,57 +10,31 @@
-    public void testAddMethod() throws Exception {
+    @Before
+    public void init() {
+        mContainerOfIntegers.add(30);
```
adding elements to container should be moved to tests methods

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -9,57 +10,31 @@
-        mContainerOfIntegers.add(28);
-        mContainerOfIntegers.add(1);
         final Integer removedItem = mContainerOfIntegers.removeByIndex(1);
         Assert.assertFalse("Container must not contain removed item any more",
```
use AssertJ for clearness

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -9,57 +10,31 @@
     }
 
-    @Test
-    public void testCopyContainer() throws Exception {
```
add test method that actually tests that copying constructor works as expected

KroosT committed:
Renamed methods.

KroosT committed:
Modified tests.

Container instance recreated before each test;
Adding elements to container moved into tests methods.
Added test method that actually tests that copying constructor works
as expected.
Used AssertJ.

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -1,37 +1,55 @@
 
+        final Container<Integer> copy = new Container<>(mContainerOfIntegers);
+
+        assertThat(copy.getList()).isNotSameAs(mContainerOfIntegers.getList());
```
what does this test check?

KroosT committed:
Fix copy constructor test.

vmatelsky commented:
LGTM

vmatelsky commented on Tasks/src/ContainerTask1/ContainerTest.java:
```diff
@@ -0,0 +1,57 @@
+        mContainerOfIntegers.removeAt(1);
+        final List<Integer> expected = Arrays.asList(30, 1, 28);
+
+        assertThat(mContainerOfIntegers.getList()).isEqualTo(expected);
```
use other method with better assertion failure message?

KroosT committed:
Used method with better assertion failure message. Packages renamed.

vmatelsky commented:
LGTM
